### PR TITLE
Fix Proxy.ListenPath stripping

### LINKED ISSUE #2421: Listen Path with regex unable to work with version as first part of path
**Do you want to request a *feature* or report a *bug*?**
BUG
**What is the current behavior?**
When listen path contains regex, api versioning with first part of path does not work
**What is the expected behavior?**
Api versioning should work
**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**
1. Import api definition
2. Trigger call to `{gateway_url}/taihoe-test/asdas234234dad/id/v1/get`

Actual:
We should get mocked response from endpoint

Testing definition:
```json
{
    "id": "5d376010cce9f316e13d6b3f",
    "name": "taihoe test",
    "slug": "taihoe-test",
    "api_id": "68e8fb04a880407c4571d4c9f96cfc53",
    "org_id": "5cc365419b7d7c07e34bcf45",
    "use_keyless": true,
    "use_oauth2": false,
    "use_openid": false,
    "openid_options": {
        "providers": [],
        "segregate_by_client": false
    },
    "oauth_meta": {
        "allowed_access_types": [],
        "allowed_authorize_types": [],
        "auth_login_redirect": ""
    },
    "auth": {
        "use_param": false,
        "param_name": "",
        "use_cookie": false,
        "cookie_name": "",
        "auth_header_name": "Authorization",
        "use_certificate": false,
        "validate_signature": false,
        "signature": {
            "algorithm": "",
            "header": "",
            "secret": "",
            "allowed_clock_skew": 0,
            "error_code": 0,
            "error_message": ""
        }
    },
    "use_basic_auth": false,
    "basic_auth": {
        "disable_caching": false,
        "cache_ttl": 0,
        "extract_from_body": false,
        "body_user_regexp": "",
        "body_password_regexp": ""
    },
    "use_mutual_tls_auth": false,
    "client_certificates": [],
    "upstream_certificates": {},
    "pinned_public_keys": {},
    "enable_jwt": false,
    "use_standard_auth": false,
    "enable_coprocess_auth": false,
    "jwt_signing_method": "",
    "jwt_source": "",
    "jwt_identity_base_field": "",
    "jwt_client_base_field": "",
    "jwt_policy_field_name": "",
    "jwt_default_policies": [],
    "jwt_issued_at_validation_skew": 0,
    "jwt_expires_at_validation_skew": 0,
    "jwt_not_before_validation_skew": 0,
    "jwt_skip_kid": false,
    "notifications": {
        "shared_secret": "",
        "oauth_on_keychange_url": ""
    },
    "enable_signature_checking": false,
    "hmac_allowed_clock_skew": -1,
    "hmac_allowed_algorithms": [],
    "base_identity_provided_by": "",
    "definition": {
        "location": "url",
        "key": "x-api-version",
        "strip_path": true
    },
    "version_data": {
        "not_versioned": false,
        "default_version": "v1",
        "versions": {
            "v1": {
                "name": "v1",
                "expires": "",
                "paths": {
                    "ignored": [],
                    "white_list": [],
                    "black_list": []
                },
                "use_extended_paths": true,
                "extended_paths": {
                    "white_list": [
                        {
                            "path": "/get",
                            "method_actions": {
                                "GET": {
                                    "action": "reply",
                                    "code": 200,
                                    "data": "{\"hello\":\"world\"}",
                                    "headers": {}
                                }
                            }
                        }
                    ],
                    "internal": []
                },
                "global_headers": {},
                "global_headers_remove": [],
                "global_size_limit": 0,
                "override_target": ""
            }
        }
    },
    "uptime_tests": {
        "check_list": [],
        "config": {
            "expire_utime_after": 0,
            "service_discovery": {
                "use_discovery_service": false,
                "query_endpoint": "",
                "use_nested_query": false,
                "parent_data_path": "",
                "data_path": "",
                "port_data_path": "",
                "target_path": "",
                "use_target_list": false,
                "cache_timeout": 60,
                "endpoint_returns_list": false
            },
            "recheck_wait": 0
        }
    },
    "proxy": {
        "preserve_host_header": false,
        "listen_path": "/taihoe-test/{test:[\\w\\d]+}/id/",
        "target_url": "http://httpbin.org/",
        "disable_strip_slash": false,
        "strip_listen_path": true,
        "enable_load_balancing": false,
        "target_list": [],
        "check_host_against_uptime_tests": false,
        "service_discovery": {
            "use_discovery_service": false,
            "query_endpoint": "",
            "use_nested_query": false,
            "parent_data_path": "",
            "data_path": "",
            "port_data_path": "",
            "target_path": "",
            "use_target_list": false,
            "cache_timeout": 0,
            "endpoint_returns_list": false
        },
        "transport": {
            "ssl_insecure_skip_verify": false,
            "ssl_ciphers": [],
            "ssl_min_version": 0,
            "proxy_url": ""
        }
    },
    "disable_rate_limit": false,
    "disable_quota": false,
    "custom_middleware": {
        "pre": [],
        "post": [],
        "post_key_auth": [],
        "auth_check": {
            "name": "",
            "path": "",
            "require_session": false
        },
        "response": [],
        "driver": "",
        "id_extractor": {
            "extract_from": "",
            "extract_with": "",
            "extractor_config": {}
        }
    },
    "custom_middleware_bundle": "",
    "cache_options": {
        "cache_timeout": 60,
        "enable_cache": true,
        "cache_all_safe_requests": false,
        "cache_response_codes": [],
        "enable_upstream_cache_control": false,
        "cache_control_ttl_header": ""
    },
    "session_lifetime": 0,
    "active": true,
    "internal": false,
    "auth_provider": {
        "name": "",
        "storage_engine": "",
        "meta": {}
    },
    "session_provider": {
        "name": "",
        "storage_engine": "",
        "meta": {}
    },
    "event_handlers": {
        "events": {}
    },
    "enable_batch_request_support": false,
    "enable_ip_whitelisting": false,
    "allowed_ips": [],
    "enable_ip_blacklisting": false,
    "blacklisted_ips": [],
    "dont_set_quota_on_create": false,
    "expire_analytics_after": 0,
    "response_processors": [],
    "CORS": {
        "enable": false,
        "allowed_origins": [],
        "allowed_methods": [],
        "allowed_headers": [],
        "exposed_headers": [],
        "allow_credentials": false,
        "max_age": 24,
        "options_passthrough": false,
        "debug": false
    },
    "domain": "",
    "certificates": [],
    "do_not_track": false,
    "tags": [
        "gcc-internet-external"
    ],
    "enable_context_vars": false,
    "config_data": {},
    "tag_headers": [],
    "global_rate_limit": {
        "rate": 0,
        "per": 0
    },
    "strip_auth_data": false
}
```

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**
2.8.3 and lower versions

### PR DESCRIPTION
**Pull request by gernest**
Fixes #2421

### CODE REVIEW

**Comment by buger** [comment]
Why we can't delegate it to gorilla mux, since it is already support regexps?

**Comment by gernest** [comment]
@buger delegating is not the issue, stripping prefix is.

Assume l`ListenPath` is  `/taihoe-test/{test:[\\w\\d]+}/id/`. If you  call  `/taihoe-test/asdas234234dad/id/v1/get`

If you need to strip ListenPath, that happens outside of gorilla/mux . Before we were just using `strings.TrimPrefix` which won't work for this case because it will treat ListenPath as tring which its not.

File: gateway/api_definition.go [hunk 1]
```diff
@@ -1313,3 +1318,18 @@ func (r *RoundRobin) WithLen(len int) int {
...
+
+var listenPathVarsRE = regexp.MustCompile(`{[^:]+(:[^}]+)?}`)
+
+func stripListenPath(listenPath, path string, muxVars map[string]string) string {
+	if !strings.Contains(listenPath, "{") {
+		return strings.TrimPrefix(path, listenPath)
+	}
+	lp := listenPathVarsRE.ReplaceAllStringFunc(listenPath, func(match string) string {
+		match = strings.TrimLeft(match, "{")
+		match = strings.TrimRight(match, "}")
```
Comment (buger) [inline gateway/api_definition.go]: We do not really care about what is on the right side of ":", so we can safely skip this line.

Comment (buger) [reply gateway/api_definition.go]: Additionally, we can just replace this both lines by: `aliasVar := strings.Split(match[1:], ":")[0]`
We know that first character is always `{`

Comment (buger) [reply gateway/api_definition.go]: Just add code comment clarifying this code.

Comment (gernest) [reply gateway/api_definition.go]: This takes into account for the case of `{key}` If we don't do that then `aliasVar="key}"`

Comment (buger) [reply gateway/api_definition.go]: I see, no worries, thats fine then

**Pull request merged**